src/js/actions/documents.js

```diff
@@ -898,18 +883,54 @@ define(function (require, exports) {
         }
 
         var newVisibility = !document.smartGuidesVisible,
-            dispatchPromise = this.dispatchAsync(events.document.GUIDES_VISIBILITY_CHANGED,
-                { documentID: document.id, smartGuidesVisible: newVisibility });
-
-        var playObject = documentLib.setSmartGuidesVisibility(newVisibility),
+            playObject = documentLib.setSmartGuidesVisibility(newVisibility),
             playPromise = descriptor.playObject(playObject);
 
-        return Promise.join(dispatchPromise, playPromise);
+        return playPromise
+            .bind(this)
+            .then(function () {
+                return _dispatchGuidesVisibilityEvent.call(this, document);
+            });
     };
     toggleSmartGuidesVisibility.action = {
         reads: [locks.JS_DOC, locks.JS_APP],
         writes: [locks.JS_DOC, locks.PS_DOC]
     };
+    
+    /**
+     * A helper function to get the visibility state of guides and smart guides from PS and dispatch
+     * GUIDES_VISIBILITY_CHANGED event to update the related menu items.
+     * 
+     * @param  {Document} document
+     * @return {Promise}
+     */
+    var _dispatchGuidesVisibilityEvent = function (document) {
+        var guidesPromise = descriptor.playObject(documentLib.getGuidesVisibility()),
+            smartGuidesPromise = descriptor.playObject(documentLib.getSmartGuidesVisibility());
+
+        return Promise.join(guidesPromise, smartGuidesPromise)
+            .bind(this)
+            .then(function (results) {
+                var guidesVisible = results[0].guidesVisibility,
+                    smartGuidesVisible = results[1].smartGuidesVisibility,
+                    guideInitPromise = Promise.resolve();
+
+                if (guidesVisible && !document.guides) {
+                    guideInitPromise = _getGuidesForDocument(document.id);
+                }
+
+                return guideInitPromise
```

Comment:
The nested function is using the `guidesVisible` `smartGuidesVisible` from the outer scope, so moving it out from the scope might not work, unless I do something like `return Promise.join(guideInitPromise, guideVisible, smartGuideVisible);` to pass the variables. What do you think?